### PR TITLE
Using MultiForkByKeyProvider for all datagen

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

File: provider/core/src/dynutil.rs [hunk 1]
```diff
@@ -213,8 +213,6 @@ macro_rules! impl_dyn_provider {
                             })
                         }
                     )+,
-                    // Don't complain if the call site has its own wildcard match
-                    #[allow(unreachable_patterns)]
```
Comment (sffc) [inline provider/core/src/dynutil.rs:217]: Request: Can you make this an error? Experiment.

File: provider/uprops/src/enum_uniset.rs [hunk 2]
```diff
@@ -99,7 +99,7 @@ impl DynProvider<UnicodePropertyV1Marker> for EnumeratedPropertyUnicodeSetDataPr
 }
 
 icu_provider::impl_dyn_provider!(EnumeratedPropertyUnicodeSetDataProvider, {
-    _ => UnicodePropertyV1Marker,
+    _k if _k.get_path().starts_with("props/") && _k.get_path().contains('=') => UnicodePropertyV1Marker,
```
Comment (sffc) [inline provider/uprops/src/enum_uniset.rs:102]: Praise: This is cool

File: provider/uprops/src/provider.rs [hunk 3]
```diff
@@ -2,57 +2,38 @@
...
 /// This data provider returns `UnicodeSet` data inside a `UnicodeProperty`
 /// data struct. The source data is in the form of a directory of TOML file(s)
 /// of data for the property(-ies) desired, as given by the ICU4C property data
 /// exporter tool.
-pub struct PropertiesDataProvider {
-    binary: BinaryPropertyUnicodeSetDataProvider,
-    enumerated: EnumeratedPropertyUnicodeSetDataProvider,
-}
-
+pub struct PropertiesDataProvider;
```
Comment (sffc) [inline provider/uprops/src/provider.rs:14]: Nit: Given a choice between naked function and fake struct, I prefer naked function.  Naming ideas:

- `create_exportable_properties_provider<T>()`
- `create_exportable_serialize_properties_provider()`

Can you return:

`eyre::Result<MultiForkByKeyProvider<Box<dyn IterableDynProvider<T>>>>`

and then parameterize the function using a `where` to make `T` valid?

File: tools/datagen/src/bin/datagen.rs [hunk 4]
```diff
@@ -269,48 +259,128 @@ fn main() -> eyre::Result<()> {
...
-            anchor1 = get_fs_exporter(&matches)?;
-            &mut anchor1
-        }
-        "blob" => {
-            anchor2 = get_blob_exporter(&matches)?;
-            &mut anchor2
-        }
-        _ => unreachable!(),
+    let all_keys = {
+        // TODO: Use central key repository
```
Comment (sffc) [inline tools/datagen/src/bin/datagen.rs:263]: Nit: Add issue number #1512

File: provider/core/src/dynutil.rs (hunk 1 above)
Comment (robertbastian) [reply provider/core/src/dynutil.rs:217]: I can't get it to work, even with `forbid`. By default it should warn anyway but it doesn't...

File: provider/uprops/src/lib.rs [hunk 5]
```diff
@@ -20,13 +20,37 @@
...
+pub use enum_codepointtrie::EnumeratedPropertyCodePointTrieProvider;
 pub use enum_uniset::EnumeratedPropertyUnicodeSetDataProvider;
+pub use script::ScriptWithExtensionsPropertyProvider;
+
+use icu_provider::fork::by_key::MultiForkByKeyProvider;
+use icu_provider::iter::IterableDynProvider;
+use icu_provider::DataMarker;
+use std::path::Path;
+
+pub fn create_exportable_properties_provider<T: DataMarker>(
```
Comment (sffc) [inline provider/uprops/src/lib.rs:38]: Nit: Same suggestion as above

File: tools/datagen/src/bin/datagen.rs [hunk 6]
```diff
@@ -337,11 +338,11 @@ fn main() -> eyre::Result<()> {
 
         Box::new(MultiForkByKeyProvider {
             providers: vec![
-                Box::new(icu_provider_cldr::create_exportable_properties_provider(
+                Box::new(icu_provider_cldr::create_exportable_provider(
                     cldr_paths.as_ref(),
                     uprops_root.clone(),
                 )?) as Box<dyn IterableDynProvider<SerializeMarker>>,
```
Comment (sffc) [inline tools/datagen/src/bin/datagen.rs:344]: Is this `as` cast necessary?

File: provider/cldr/src/lib.rs [hunk 7]
```diff
@@ -50,9 +50,12 @@ use transform::locale_canonicalizer::likely_subtags::LikelySubtagsProvider;
 use transform::plurals::PluralsProvider;
 use transform::time_zones::TimeZonesProvider;
 
-pub fn create_exportable_properties_provider<T: DataMarker>(
+#[cfg(not(feature = "icu_list"))]
+type ListProvider = PluralsProvider; // we can't cfg-exclude part of the bound, but we can do this...
```
Comment (sffc) [inline provider/cldr/src/lib.rs:54]: I approve of this hack.  Hopefully we can remove the `cfg` soon, once BurntSushi delivers 0.2...

File: tools/datagen/src/bin/datagen.rs (hunk 6 above)
Comment (robertbastian) [reply tools/datagen/src/bin/datagen.rs:344]: Yes

**Review by sffc — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]
wonderful!

**Pull request merged**